packages/labs/ssr/src/lib/element-renderer.ts

```diff
@@ -42,15 +42,19 @@ export const getElementRenderer = (
   return new FallbackRenderer(tagName);
 };
 
-export interface ShadowRootOptions {
-  mode: 'open' | 'closed';
-  delegatesFocus?: boolean;
-}
+// TODO (justinfagnani): remove in favor of ShadowRootInit
+/**
+ * @deprecated Use ShadowRootInit instead
+ */
+export type ShadowRootOptions = ShadowRootInit;
 
 /**
  * An object that renders elements of a certain type.
  */
 export abstract class ElementRenderer {
+  // TODO (justinfagnani): We shouldn't assume that ElementRenderer subclasses
+  // create an element instance. Move this to a base class for renderes that
```

Comment:
What are you using `.element` for? I think renderers without instantiated elements should be as capable as those with, and the renderer is supposed to abstract away whether there's a real instance and how behavior, state and rendering are managed.

Reply:
Doing `if (key in renderer.element)` for whether to set property or attribute and to get `renderer.element.attributes` after calling `renderer.connectedCallback()` for any reflected attributes.

Reply:
Two thoughts:

1. It's ok for a Lit-specific integration to use the LitElementRenderer interface. Though we would like a non-Lit-specific Integration.
2. I think we should add the needed API to ElementRenderer. If you can set properties and attributes via the renderer, and we also need to query things about properties, I don't see how it doesn't violate the contract for `setProperty()` to directly do an `in` check.

    IOW, if it's ok to do `x in renderer.element` why isn't it ok to do `renderer.element.x = value`? If the renderer is a proxy for the element, we need to have API for everything we might need to do to the element. Otherwise renderers that don't have an underlying element will eventually need to make an actual proxy-ing object that implements the parts of the `HTMLElement` API that integrations are using.

> `renderer.element.attributes` after calling `renderer.connectedCallback()` for any reflected attributes

Why do you need to do this? `renderAttributes()` should render reflected attributes. Do you need them for a reason other than rendering, or do you need them represented in the vdom?

Reply:
For the React integration I'm letting React render the host element and therefore need to provide React's createElement the attributes as props.